### PR TITLE
Introduce new `RecordingStream` logging APIs (sunset `MsgSender`)

### LINKED ISSUE #2825: The end of `MsgSender` and streamlining logging APIs across all SDKs
I've been talking about the retirement of `MsgSender` for a long time but never made an actual issue, apparently.

`MsgSender` is heavy, clunky and almost entirely superfluous now that `RecordingStream` handles all the hard parts of logging for all 3 SDKs.
With the new archetype-based API, it will somehow become even more superfluous than it already is.

The removal of `MsgSender` is also an opportunity to streamline the behavior of all 3 SDKs behind 2 entrypoints:
- `log(entity_path, archetype)` (aliased to `log_archetype()`), which is what I like to call the high-level object-oriented API.
- `log_components(entity_path, components...)`, which is what I like to call the low-level ECS API.

### Example of using the high-level API

As a user, the high-level API gives you guarantees that things will _just work_ :tm:.

You don't need to worry about how things compose together, rather you think in more familiar terms: objects.

In Rust:
```rust
// logging a Points3D archetype, both the timepoint and the recordingstream are picked up from TLS:
rerun::log("points", &Points3D::new([(0.0, 0.0, 0.0), (1.0, 1.0, 1.0)])
    .with_radii([1.0, 2.0])
    .with_colors([0xFF0000]))?;

// which is the same as:
rerun::log_archetype("points", &Points3D::new([(0.0, 0.0, 0.0), (1.0, 1.0, 1.0)])
    .with_radii([1.0, 2.0])
    .with_colors([0xFF0000]))?;

// we can also imagine exposing one with an explicit recordingstream argument:
rerun::log2(&rec_stream, "points", &Points3D::new([(0.0, 0.0, 0.0), (1.0, 1.0, 1.0)])
    .with_radii([1.0, 2.0])
    .with_colors([0xFF0000]))?;
```

And in Python:
```python
# logging a Points3D archetype, both the timepoint and the recordingstream are picked up from TLS:
rr.log("simple", rr.Points3D([[0, 0, 0], [1, 1, 1]], radii=[1, 2], colors=0xFF0000))

# which is the same as:
rr.log_archetype("simple", rr.Points3D([[0, 0, 0], [1, 1, 1]], radii=[1, 2], colors=0xFF0000))
```

And probably something very similar in C++, according to my recent discussions with @Wumpf!

### Example of using the low-level API

As a user, the low-level API gives you access to all of Rerun's power, but comes at a cost: you need to think in terms of composition, and understand all the rules involved (partial updates, joins, instance keys, archetype definitions...).
And of course it's much more verbose and much less friendly (can't autocomplete when we don't know what you're looking for to begin with!).

This allows you to incrementally build up archetypes over multiple timestamps, to do partial updates, to build custom views, to use extension components, etc.

In Rust:
```rust
// logging that same 3D point cloud but using composition this time:
rerun::log_components(
    "points",
    [rrc::Point3D::new(0.0, 0.0, 0.0), rrc::Point3D::new(1.0, 1.0, 1.0)],
    [rrc::Radius::new(0.0), rrc::Radius::new(1.0)],
    [rrc::Color::new(0xFF0000)],
)?;

// or, expressed in a nicer way ():
rerun::log_components(
    "points",
    [(0.0, 0.0, 0.0), (1.0, 1.0, 1.0)].map(rrc::Point3D::from),
    [0.0, 1.0].map(rrc::Radius::from),
    [0xFF0000].map(rrc::Color::from),
)?;
```

Python:
```python
# logging that same 3D point cloud but using composition this time:
rr.log_components(
    "simple",
    rrc.Point3DArray([[0, 0, 0], [1, 1, 1]]),
    rrc.RadiusArray([1, 2]),
    rrc.ColorArray([0xFF0000]),
)
```

And probably something very similar in C++.

### PR DESCRIPTION
**Pull request by teh-cmc**
**Commit by commit**

Implements the new logging APIs, which supersede the ol' `MsgSender`:
```rust
pub fn log(
    &self,
    ent_path: impl Into<EntityPath>,
    arch: &impl Archetype,
) -> RecordingStreamResult<()>;

pub fn log_archetype(
    &self,
    ent_path: impl Into<EntityPath>,
    timeless: bool,
    arch: &impl Archetype,
) -> RecordingStreamResult<()>;

pub fn log_component_lists<'a>(
    &self,
    ent_path: impl Into<EntityPath>,
    timeless: bool,
    num_instances: u32,
    comp_lists: impl IntoIterator<Item = &'a dyn ComponentList>,
) -> RecordingStreamResult<()>;
```

This PR also does _a lot_ of test/example clean up, in particular the doc examples which were broken in many, many ways.
There is still a lot to be done on that front for `0.9`, which I'll describe in a future ticket.

---

Example 1: Logging an archetype
```rust
rec.log("points", &Points3D::new([(0.0, 0.0, 0.0), (1.0, 1.0, 1.0)]))?;
```

Example 2: Logging a component batch
```rust
rec.log_component_lists(
    "bounds",
    false,
    1,
    [&Rect2D::XCYCWH(Vec4D([3.0, 1.5, 8.0, 9.0]).into()) as _],
)?;
```

Fixes #2825 

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3184) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3184)
- [Docs preview](https://rerun.io/preview/bc79e20b98cf5625bba071d1a1ef4ed17c2e2b59/docs) 
- [Examples preview](https://rerun.io/preview/bc79e20b98cf5625bba071d1a1ef4ed17c2e2b59/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

**Comment by emilk** [comment]
> ```rust
> pub fn log(
>     &self,
>     ent_path: impl Into<EntityPath>,
>     arch: &impl Archetype,
> ) -> RecordingStreamResult<()>;
> 
> pub fn log_archetype(
>     &self,
>     ent_path: impl Into<EntityPath>,
>     timeless: bool,
>     arch: &impl Archetype,
> ) -> RecordingStreamResult<()>;
> ```

Without having seen anything else, I assume `log` just calls `log_archetype` with `timeless=false`?

It's a bit weird that the second function has the `_archetype` suffix when that is not its distinguishing difference (they both log archetypes).

I'd suggest `log_timeless(ent_path, arch)` or `log_with_timeless(ent_path, bool, arch)`

File: crates/re_sdk/src/recording_stream.rs [hunk 1]
```diff
@@ -595,12 +595,75 @@ impl RecordingStream {
...
+        for cell in cells {
+            if num_instances > 1 && cell.num_instances() == 1 {
+                splatted.push(cell);
+            } else {
+                instanced.push(cell);
+            }
+        }
+
+        // TODO: not sure about this though
+        let timepoint = TimePoint::default();
```
Comment (jleibs) [inline crates/re_sdk/src/recording_stream.rs]: Shouldn't this be:
```
TimePoint::timeless()
```

**Comment by teh-cmc** [comment]
`log_archetype` -> `log_timeless` :+1:

**Comment by teh-cmc** [comment]
All failing checks are due to Github's servers being down, merging.

**Pull request merged**